lib/rspec/support/object_formatter.rb

```diff
@@ -65,30 +66,44 @@ def prepare_for_inspection(object)
       end
 
       def prepare_array(array)
-        array.map do |element|
-          if element.equal?(array)
-            InspectableItem.new('[...]')
-          else
-            prepare_for_inspection(element)
-          end
+        with_entering_structure(array) do
+          array.map { |element| prepare_element(element) }
         end
       end
 
       def prepare_hash(input_hash)
-        input_hash.inject({}) do |output_hash, key_and_value|
-          key, value = key_and_value.map do |element|
-            if element.equal?(input_hash)
-              InspectableItem.new('{...}')
-            else
-              prepare_for_inspection(element)
-            end
+        with_entering_structure(input_hash) do
+          input_hash.inject({}) do |output_hash, key_and_value|
+            key, value = key_and_value.map { |element| prepare_element(element) }
+            output_hash[key] = value
+            output_hash
           end
+        end
+      end
 
-          output_hash[key] = value
-          output_hash
+      def prepare_element(element)
+        if recursive_structure?(element)
+          case element
+          when Array then InspectableItem.new('[...]')
+          when Hash then InspectableItem.new('{...}')
+          else raise
```

Comment:
Actually the `raise` is a sort of assertion and will never be run since `#recursive_structure?` returns true only when a hash or array is given.